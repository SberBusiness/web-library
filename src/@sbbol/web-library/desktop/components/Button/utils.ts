import * as React from 'react';
import {EButtonTheme, EButtonSize} from '@sbbol/web-library/desktop/components/Button/enums';

/**
 * Не происходит фокус при клике на button в сафари и в мозиле на маках.
 * https://developer.mozilla.org/en-US/docs/Web/HTML/Element/button#Clicking_and_focus
 *
 * Данная функция должна вызываться только синхронно. (Не будет работать в колбэках setState)
 * https://stackoverflow.com/questions/42089795/reactjs-cant-set-state-from-an-event-with-event-persist
 */
export const focusButton = (event: React.MouseEvent<HTMLButtonElement>): void => {
    event.currentTarget?.focus();
};

/**
 * Мапа энама типа кнопки на стилевой класс.
 */
export const mapButtonThemeToCssClass = {
    [EButtonTheme.GENERAL]: 'cssClass[buttonGeneral]',
    [EButtonTheme.SECONDARY]: 'cssClass[buttonSecondary]',
    [EButtonTheme.DANGER]: 'cssClass[buttonDanger]',
    [EButtonTheme.DOTS]: 'cssClass[buttonDots]',
    [EButtonTheme.LINK]: 'cssClass[buttonLink]',
};

/**
 * Мапа энама размера кнопки на стилевой класс.
 */
export const mapButtonSizeToCssClass = {
    [EButtonSize.MD]: 'cssClass[buttonMd]',
    [EButtonSize.SM]: 'cssClass[buttonSm]',
};
