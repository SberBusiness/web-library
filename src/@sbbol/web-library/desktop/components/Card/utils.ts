import {ECardRoundingSize} from '@sbbol/web-library/desktop/components/Card/enums';

/** Соответствие типа скругления карточки стилевому классу. */
export const mapCardRoundingSizeToCssClass = {
    [ECardRoundingSize.MD]: 'cssClass[roundingMD]',
    [ECardRoundingSize.SM]: 'cssClass[roundingSM]',
};
