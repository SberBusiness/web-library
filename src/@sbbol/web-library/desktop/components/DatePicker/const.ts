/** Формат даты, отображаемый при вооде. */
export const inputDateFormat = 'DD.MM.YYYY';

/** Стандартное значение плейсхолдера. */
export const defaultDatePlaceholder = 'дд.мм.гггг';
