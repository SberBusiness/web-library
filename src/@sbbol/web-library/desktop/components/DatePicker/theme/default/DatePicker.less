@import '../../../../common/theme/default';

.datePicker {
    position: relative;
    width: @picker-width;

    .globalCalendar {
        .pickerCalendarPopover();
    }

    &.topOrientation {
        .globalCalendar {
            bottom: 36px;
            margin-top: 0;
        }
    }
}
