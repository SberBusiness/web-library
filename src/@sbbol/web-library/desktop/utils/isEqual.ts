export const isEqual = (a: any, b: any): boolean => {
    if (a === b) {
        return true;
    }

    if (a && b && typeof a === 'object' && typeof b === 'object') {
        if (a.constructor !== b.constructor) {
            return false;
        }

        let length: number;
        let i: number;
        let keys: string[];
        if (Array.isArray(a) && Array.isArray(b)) {
            length = a.length;
            if (length !== b.length) {
                return false;
            }
            for (i = length; i-- !== 0; ) {
                if (!isEqual(a[i], b[i])) {
                    return false;
                }
            }
            return true;
        }

        if (a instanceof Map && b instanceof Map) {
            if (a.size !== b.size) {
                return false;
            }
            for (const e of a.entries()) {
                if (!b.has(e[0])) {
                    return false;
                }
            }
            for (const e of a.entries()) {
                if (!isEqual(e[1], b.get(e[0]))) {
                    return false;
                }
            }
            return true;
        }

        if (a instanceof Set && b instanceof Set) {
            if (a.size !== b.size) {
                return false;
            }
            for (const e of a.entries()) {
                if (!b.has(e[0])) {
                    return false;
                }
            }
            return true;
        }

        if (a.constructor === RegExp) {
            return a.source === b.source && a.flags === b.flags;
        }

        if (a.valueOf !== Object.prototype.valueOf) {
            return a.valueOf() === b.valueOf();
        }

        if (a.toString !== Object.prototype.toString) {
            return a.toString() === b.toString();
        }

        keys = Object.keys(a);
        length = keys.length;
        if (length !== Object.keys(b).length) {
            return false;
        }

        for (i = length; i-- !== 0; ) {
            if (!Object.prototype.hasOwnProperty.call(b, keys[i])) {
                return false;
            }
        }

        for (i = length; i-- !== 0; ) {
            const key = keys[i];
            if (key === '_owner' && a.$$typeof) {
                // Не сравниваем родителей в React.
                continue;
            }
            if (!isEqual(a[key], b[key])) {
                return false;
            }
        }

        return true;
    }

    return a !== a && b !== b; // Вернёт true если оба NaN.
};
