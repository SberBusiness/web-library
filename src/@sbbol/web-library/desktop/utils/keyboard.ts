/**
 * Коды клавиш.
 */
export const EVENT_KEYS = {
    ARROW_DOWN: ['ArrowDown', 'Down'],
    ARROW_LEFT: ['ArrowLeft', 'Left'],
    ARROW_RIGHT: ['ArrowRight', 'Right'],
    ARROW_UP: ['ArrowUp', 'Up'],
    BACKSPACE: ['Backspace'],
    DELETE: ['Delete', 'Del'],
    ENTER: ['Enter'],
    ESC: ['Escape', 'Esc'],
    SPACE: ['Spacebar', 'Space'],
    TAB: ['Tab'],
    X: ['KeyX', 'X', 'x'],
    Y: ['KeyY', 'Y', 'y'],
    Z: ['KeyZ', 'Z', 'z'],
};

// Тип ключи EVENT_KEYS.
export type TEventKeys = keyof typeof EVENT_KEYS;

// Тип значения EVENT_KEYS.
export type TEventKeysValues = typeof EVENT_KEYS[TEventKeys];

/**
 * Числовые коды клавиш.
 */
export const EVENT_KEY_CODES = {
    ARROW_DOWN: 40,
    ARROW_LEFT: 37,
    ARROW_RIGHT: 39,
    ARROW_UP: 38,
    ENTER: 13,
    ESC: 27,
    SPACE: 32,
    TAB: 9,
};

/**
 * Предикат для проверки нажатой клавиши на соответствие с target. isKey(key, 'Enter');
 * @param {string | number} key Код нажатой клавиши.
 * @param {keyof typeof EVENT_KEY_CODES} target Ожидаемая клавиша.
 */
export function isKey(key: string | number, target: keyof typeof EVENT_KEY_CODES): boolean {
    if (typeof key === 'string') {
        return EVENT_KEYS[target].includes(key);
    }
    // IE fallback.
    return key === EVENT_KEY_CODES[target];
}
