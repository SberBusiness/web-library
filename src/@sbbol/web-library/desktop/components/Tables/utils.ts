import {EScreenSize} from '@sbbol/web-library/common/enums/EScreenSize';
import {ECellType, EHorizontalAlign, EVerticalAlign} from '@sbbol/web-library/desktop/components/Tables/TableBasic/enums';
import {cssClass} from '@sbbol/web-library/desktop/utils/cssClass';

/**
 * Преобразование значения горизонтального выравнивания в css-класс.
 * @param {EHorizontalAlign} [horizontalAlign] Значение для маппинга.
 */
export const mapHorizontalAlignToClassName = (horizontalAlign: EHorizontalAlign = EHorizontalAlign.LEFT): string => {
    switch (horizontalAlign) {
        case EHorizontalAlign.LEFT:
            return 'cssClass[alignLeft]';
        case EHorizontalAlign.RIGHT:
            return 'cssClass[alignRight]';
        case EHorizontalAlign.CENTER:
            return 'cssClass[alignCenter]';
    }
};

/**
 * Преобразование значения вертикального выравнивания в css-класс.
 * @param {EVerticalAlign} [verticalAlign] Значение для маппинга.
 */
export const mapVerticalAlignToClassName = (verticalAlign: EVerticalAlign = EVerticalAlign.BASELINE): string => {
    switch (verticalAlign) {
        case EVerticalAlign.BASELINE:
            return 'cssClass[verticalAlignBaseline]';
        case EVerticalAlign.SUB:
            return 'cssClass[verticalAlignSub]';
        case EVerticalAlign.SUPER:
            return 'cssClass[verticalAlignSuper]';
        case EVerticalAlign.TEXT_TOP:
            return 'cssClass[verticalAlignTextTop]';
        case EVerticalAlign.TEXT_BOTTOM:
            return 'cssClass[verticalAlignTextBottom]';
        case EVerticalAlign.MIDDLE:
            return 'cssClass[verticalAlignMiddle]';
        case EVerticalAlign.TOP:
            return 'cssClass[verticalAlignTop]';
        case EVerticalAlign.BOTTOM:
            return 'cssClass[verticalAlignBottom]';
    }
};

/**
 * Преобразование значений скрытия колонок в css-классы.
 * @param {EScreenSize[]} [hideScreenSizes] Массив размеров ширины экрана, при которых скрывать колонку.
 */
export const mapHideScreenSizesToClassName = (hideScreenSizes?: EScreenSize[]): {[p: string]: boolean}[] | undefined =>
    hideScreenSizes && Array.isArray(hideScreenSizes)
        ? hideScreenSizes.map((size) => ({
              [cssClass(`hide${size}`)]: true,
          }))
        : undefined;

/**
 * Преобразование типа ячейки в css-класс.
 * @param {ECellType} [cellType] Тип ячейки.
 */
export const mapCellTypeToClassName = (cellType: ECellType = ECellType.TEXT): string => cssClass(`${cellType}Type`);
