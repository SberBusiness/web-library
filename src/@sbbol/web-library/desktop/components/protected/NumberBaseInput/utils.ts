import {amountComparator, calculateAmountMaxValue, formatAmount} from '@sbbol/web-library/desktop/utils/amountUtils';
import {AmountConst} from '@sbbol/web-library/desktop/common/consts/AmountConst';

/** Формат поля ввода суммы. */
const amountRegExp = new RegExp(`^([1-9][0-9]*|[0-9])(${AmountConst.DecimalSeparator}[0-9]*)?$`);

/**
 * Проверка на максимальное значение.
 *
 * @param {string} value Значение.
 * @param {string | number | undefined} max Максимальное значение числа в поле.
 * @param {number} maxLength Максимальная длина поля.
 * @param {number} fractionLength Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
 */
const checkMaxValue = (value: string, max: string | number | undefined, maxLength: number, fractionLength: number): boolean => {
    if (!amountRegExp.test(value)) {
        return false;
    }

    const resultMaxValue = getMaxValue(max, maxLength, fractionLength);
    if (!resultMaxValue) {
        return false;
    }

    return amountComparator(value, resultMaxValue) <= 0;
};

/**
 * Вычисляет максимальное значение для поля суммы.
 *
 * @param {string | number | undefined} max Максимальное значение числа в поле.
 * @param {number} maxLength Максимальная длина поля.
 * @param {number} fractionLength Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
 * */
const getMaxValue = (max: string | number | undefined, maxLength: number, fractionLength: number): string | undefined => {
    let maxValue: number;
    if (typeof max === 'string') {
        maxValue = Number.parseFloat(max);
    } else if (typeof max === 'number') {
        maxValue = max;
    }
    if (!Number.isInteger(maxValue!) && maxLength !== undefined) {
        return calculateAmountMaxValue(maxLength, fractionLength);
    } else if (Number.isInteger(max) && maxLength === undefined) {
        return String(max);
    } else {
        const maxAmountValueByLength = calculateAmountMaxValue(maxLength, fractionLength);
        const maxAsString = String(max);
        // проверить входит ли максимальное число в диапазон ограниченный длиной
        if (amountComparator(maxAmountValueByLength as string, maxAsString) < 0) {
            // если нет, ограничить максимальной длиной
            return maxAmountValueByLength;
        } else {
            // если да - максимальным значением
            return maxAsString;
        }
    }
};

/**
 * Вычисляем позицию каретки. (скопирована из СББОЛ)
 *
 * @param {string} prevValueFormatted Прошлое отформатированное значение.
 * @param {string} newValueRaw Новое не отформатированное значение.
 * @param {number} fractionLength Количество знаков в дробной части. При значении 0 подразумевается целочисленное значение.
 * @param {number} currentCaretPosition Позиция каретки в поле ввода.
 */
const countCaret = (prevValueFormatted: string, newValueRaw: string, fractionLength: number, currentCaretPosition: number): number => {
    const spacesInOldValue = (prevValueFormatted.match(/\s/g) || []).length;
    const spacesInNewValue = (formatAmount(newValueRaw, fractionLength).match(/\s/g) || []).length;
    let caretSpaceOffset = spacesInNewValue - spacesInOldValue;

    if (prevValueFormatted.length - newValueRaw.length > 1) {
        // Если было удалено больше одного символа. Делаем reset caretSpaceOffset.
        caretSpaceOffset = 0;
    } else if (newValueRaw.substr(0, 1) === AmountConst.GroupsSeparator) {
        // Если впереди пробел (" 786 673") сдвигаем каретку.
        caretSpaceOffset++;
    }

    return currentCaretPosition + caretSpaceOffset;
};

export {checkMaxValue, countCaret};
