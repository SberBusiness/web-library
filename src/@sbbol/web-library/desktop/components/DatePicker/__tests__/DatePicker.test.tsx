import React from 'react';
import moment from 'moment';
import renderer from 'react-test-renderer';
import {getFormattedDate} from '@sbbol/web-library/desktop/utils/dateUtils';
import {DatePicker} from '../DatePicker';
import {allure} from '@jest/unit/allure-report';

jest.mock('@sberbusiness/icons/CalendarSrvIcon20', () => ({
    CalendarSrvIcon20: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretleftSrvxIcon24', () => ({
    CaretleftSrvxIcon24: 'svg',
}));

jest.mock('@sberbusiness/icons/CaretrightSrvxIcon24', () => ({
    CaretrightSrvxIcon24: 'svg',
}));

describe('DatePicker', () => {
    beforeEach(() => {
        allure.feature('DatePicker');
    });

    (global as any).document = {
        addEventListener: jest.fn(),
    };

    it('renders correctly', () => {
        const handleChange = jest.fn();
        const value = getFormattedDate(moment('20210101'));
        const tree = renderer.create(<DatePicker value={value} onChange={handleChange} />).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
