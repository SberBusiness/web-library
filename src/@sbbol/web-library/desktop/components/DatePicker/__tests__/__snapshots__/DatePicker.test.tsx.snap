// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatePicker renders correctly 1`] = `
<div
  className="cssClass[datePicker]"
>
  <div
    className="cssClass[inputBlock]"
  >
    <div
      className="cssClass[wrapper]"
    >
      <input
        className="cssClass[placeholder]"
        readOnly={true}
        tabIndex={-1}
        value="01.01.2021"
      />
      <input
        autoComplete="off"
        className="cssClass[input]"
        maxLength={10}
        onBlur={[Function]}
        onChange={[Function]}
        onCut={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyPress={[Function]}
        onPaste={[Function]}
        spellCheck={false}
        value="01.01.2021"
      />
    </div>
    <span
      className="cssClass[calendarIcon]"
      onClick={[Function]}
    >
      <svg />
    </span>
  </div>
</div>
`;
